rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+   
+   Mutable memory is similarly made non-deterministic, discussed later in
+   [havociing](#memory-predicates-and-havocing).
+
+   The expanded version is equivalent to this:
+  
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let dividend = v.len() - 1;
+     let divisor = 3;
+     kani::assert(divisor != 0);
+     let result = kani::any();
+     kani::assume(result <= dividend);
+     let some_idx = result;
+     v[some_idx];
+   }
+   ```
+
+   Notice that this performs no actual computation for `my_div` (other than the
+   conditions) which allows us to avoid something potentially costly.
+
+Also notice that Kani was able to express both contract checking and stubbing
+with existing capabilities; the important feature is the enforcement. The
+checking is, by construction, performed **against the same condition** that is
+later used as stub, which ensures soundness (see discussion on lingering threats
+to soundness in the [future](#future-possibilities) section) and guarding against
+stubs diverging from their checks.
+
+### History Variables
+
+Kani's contract language contains additional support for reasoning about changes
+to memory. One case where this is necessary is whenever `ensures` needs to
+reason about state before the function call. By default it only has access to
+state after the call completes, which will be different if the call mutates
+memory.
+
+Consider the `Vec::pop` function
+
+```rs
+impl<T> Vec<T> {
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+If we want to describe in which case the result is `Some`, we need to know
+whether `self` is empty *before* `pop` is called. To do this Kani provides the
+`old(EXPR)` pseudo function, which evaluates `EXPR` before the call (e.g. to
+`pop`) and makes the result available to `ensures`. It is used like so:
+
+```rs
+impl<T> Vec<T> {
+  #[kani::ensures(old(self.is_empty()) || result.is_some())]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+`old` allows evaluating any (side-effect free[^side-effects]) Rust expression.
+The bowwor checker enforces the result of `old` cannot observe the mutations
+from e.g. `pop`, as that would defeat the purpose. If `your` expression in `old`
+returns borrowed content, make a copy instead (using e.g. `clone()`).
+
+Note also that `old` is syntax, not a function and implemented as an extraction
+and lifting during code generation. It can reference e.g. `pop`'s arguments but
+not local variables. Compare the following
+
+**Invalid ❌:** `#[kani::ensures({ let x = self.is_empty(); old(x) } || result.is_some())]`</br>
+**Valid ✅:** `#[kani::ensures(old({ let x = self.is_empty(); x }) || result.is_some())]`
+
+And it will only be recognized as `old(...)`, not as `let old1 = old; old1(...)` etc.
+
+### Memory Predicates and Havocing
+
+The last new feature added are predicates to refine a function's access to heap
```

Comment:
I would prefer if we restructure this section.
1. Talk about function side effects at the top. Mention that the last part of a function specification is to determine what kind of side effects this function can trigger. Mention mutable static variables, mutable references and interior mutability.
2. Mention that we will add `#[kani::assign]` to allow users to specify which side effects a function produces.
3. Mention that Kani will assume that all mutable references and interior mutable elements can be modified, unless there's an existing assigns clause that narrows down the scope of the modifications.
4. Mention that changes to mutable static variables must be explicitly specified.